### PR TITLE
Feature/New CLI command "list package-ids"

### PR DESCRIPTION
**Pull request by franramirez688**
Changelog: omit
Docs: omit 

```
$ conan list package-ids -h
usage: conan list package-ids [-h] [-f {cli,json}] [-r REMOTE | -a] [-c] reference

positional arguments:
  reference

optional arguments:
  -h, --help            show this help message and exit
  -f {cli,json}, --format {cli,json}
                        Select the output format: cli, json. 'cli' is the default output.
  -r REMOTE, --remote REMOTE
                        Name of the remote to add
  -a, --all-remotes     Search in all the remotes
  -c, --cache           Search in the local cache
```
Docs: https://github.com/conan-io/docs/pull/XXXX

Depends on: https://github.com/conan-io/conan/pull/9288

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/cli/commands/list.py [hunk 1]
```diff
@@ -1,67 +1,83 @@
...
-                recipe_color
-            )
+
+def list_package_revisions_cli_formatter(results):
+    _list_revisions_cli_formatter(results, "packages")
 
 
+def list_package_ids_cli_formatter(results):
+    # Artifactory uses field 'requires', conan_center 'full_requires'
+    requires_fields = ("requires", "full_requires")
```
Comment (memsharded) [inline conans/cli/commands/list.py:61]: This will be changed too withe new package_id, I know the ``full_requires`` thing is weird.

File: conans/client/api/conan_api.py [hunk 2]
```diff
@@ -239,39 +241,120 @@ def search_remote_recipes(self, query, remote):
...
+                  "reference": "libcurl/7.77.0#2a9c4fcc8d76d891e4db529efbe24242",
+                  "results": {"d5f16437dd4989cc688211b95c24525589acaafd": {
+                                    "settings": {"compiler": "apple-clang",...},
+                                    "options": {'options': {'shared': 'False',...}},
+                                    "requires": [
+                                        'mylib/1.0.8:3df6ebb8a308d309e882b21988fd9ea103560e16',...]}
+                            }
+                 }
+        """
+        ref = ConanFileReference.loads(reference)
```
Comment (memsharded) [inline conans/client/api/conan_api.py:356]: When this gets too heavy on business logic, we should delegate the code to helper files.
No need to change this here, we will do it when we check the whole api.

File: conans/test/integration/command_v2/list_package_ids_test.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    def _upload_recipe(self, remote, reference):
+        self.client.save({'conanfile.py': GenConanfile()})
+        self.client.run("export . {}".format(reference))
+        self.client.run("upload --force -r {} {}".format(remote, reference))
+
+
+class TestParams(TestListPackageIdsBase):
+    def test_fail_if_reference_is_not_correct(self):
+        self.client.run("list package-ids whatever", assert_error=True)
+        assert "ERROR: Specify the 'name' and the 'version'" in self.client.out
```
Comment (memsharded) [inline conans/test/integration/command_v2/list_package_ids_test.py:51]: We need to improve the UX (not necessarily here). Error message should probably be on the form:

``"whatever" is not a valid reference, provide a reference in the form "name/version"``

File: conans/test/integration/command_v2/list_package_ids_test.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        remote_name = "remote1"
+        recipe_name = "test_recipe/1.0.0@user/channel"
+        self._add_remote(remote_name)
+        self._upload_recipe(remote_name, recipe_name)
+
+        self.client.run("list package-ids -r remote1 test_recipe/1.0.0")
+
+        expected_output = (
+            "remote1:\n"
+            "  There are no matching recipes\n"
```
Comment (memsharded) [inline conans/test/integration/command_v2/list_package_ids_test.py]: I cannot find a test with recipe but without binaries: What is the output in that case?

**Review by memsharded — APPROVED** [review]
This is good enough too, can be merged.

There is some engineering work to do, but that can wait until functionality stabilizes a bit more.

File: conans/test/integration/command_v2/list_package_ids_test.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        self.client.run("create . {}".format(reference))
+
+    def _upload_recipe(self, remote, reference):
+        self.client.save({'conanfile.py': GenConanfile()})
+        self.client.run("export . {}".format(reference))
+        self.client.run("upload --force -r {} {}".format(remote, reference))
+
+
+class TestParams(TestListPackageIdsBase):
+    def test_fail_if_reference_is_not_correct(self):
```
Comment (czoido) [inline conans/test/integration/command_v2/list_package_ids_test.py:49]: We are not testing a reference that does not exist in cache but has a valid format, like:
`conan list package-ids "notexisting/1.0@user/channel#whatever_not_existing"`
This is not returning an error.

File: conans/cli/commands/list.py [hunk 6]
```diff
@@ -1,67 +1,83 @@
...
+    general_fields = ("options", "settings")
+
+    for remote_results in results:
+        _print_common_list_cli_output(remote_results, "references")
+        reference = remote_results["reference"]
+        for pkg_id, props in remote_results["results"].items():
+            cli_out_write(repr(PackageReference(reference, pkg_id)), fg=reference_color, indentation=2)
+            for prop_name, values in props.items():
+                if prop_name in requires_fields:
```
Comment (czoido) [inline conans/cli/commands/list.py]: I think we should check here if `values` is empty or not so that it does not print the title like `requires:` if there are no requires.

File: conans/test/integration/command_v2/list_package_ids_test.py (hunk 3 above)
Comment (franramirez688) [reply conans/test/integration/command_v2/list_package_ids_test.py:51]: Agree. This is coming from `ref.py` so it should go in another PR 😉

File: conans/test/integration/command_v2/list_package_ids_test.py (hunk 5 above)
Comment (franramirez688) [reply conans/test/integration/command_v2/list_package_ids_test.py:49]: @czoido we're doing that plenty of times along the test suite. For example, `test_by_default_search_only_in_cache` is checking:

```
        Local Cache:
          There are no matching references
```
So it should be a valid test 😉  This test is done for cache and remote as well.

Comment (czoido) [reply conans/test/integration/command_v2/list_package_ids_test.py:49]: That test is failing in the PR: https://ci.conan.io/blue/organizations/jenkins/ConanTestSuitev2/detail/PR-9310/2/pipeline/34
Anyway, I think if you want to check the package-id's of a specific recipe revision that does not exist in the cache, we should raise an error not returning that there are not matching recipes but saying that that recipe revision is not in the cache.

File: conans/client/api/conan_api.py [hunk 7]
```diff
@@ -268,23 +269,100 @@ def _get_revisions(self, ref, remote=None):
...
+            try:
+                packages_props = self.app.remote_manager.search_packages(remote, rrev, None)
+            except NotFoundException:
+                # This exception must be catched manually due to a server inconsistency:
+                # Artifactory API returns an empty result if the recipe doesn't exist, but
+                # Conan Server returns a 404. This probably should be fixed server side,
+                # but in the meantime we must handle it here
+                packages_props = {}
+        else:
+            rrev = ref if ref.revision else self.app.cache.get_latest_rrev(remote, ref)
```
Comment (czoido) [inline conans/client/api/conan_api.py]: This method does not have the remote argument.

**Review by czoido — CHANGES_REQUESTED** [review]
I think we are missing some tests, check how this one fails:
```python
def test_existing_reference(self):
    self.client.save({"conanfile.py": GenConanfile().with_name("pkg").with_version("1.0")})
    self.client.run(f"create .")
    self.client.run(f"list package-ids pkg/1.0")
```

File: conans/client/api/conan_api.py (hunk 7 above)
Comment (franramirez688) [reply conans/client/api/conan_api.py]: Completely right. The thing is that I have a test for cache and remote cases but as you said I'm missing something... Thanks for the catch @czoido

File: conans/test/integration/command_v2/list_package_ids_test.py (hunk 5 above)
Comment (franramirez688) [reply conans/test/integration/command_v2/list_package_ids_test.py:49]: I tend to say that this check could go in another PR because it should be something that affects to all the list context, not only the `package-ids` subcommand

**Review by memsharded — APPROVED** [review]

File: conans/cli/commands/list.py [hunk 8]
```diff
@@ -40,7 +41,7 @@ def list_recipes_cli_formatter(results):
 def _list_revisions_cli_formatter(results, ref_type):
     for remote_results in results:
         _print_common_list_cli_output(remote_results, ref_type)
```
Comment (tapia) [inline conans/cli/commands/list.py]: This function's name (`_print_common_list_cli_output`) is not very descriptive. How about `_print_remote_header()` or something like that?

Comment (franramirez688) [reply conans/cli/commands/list.py]: That function is printing the `Local Cache` line as well so I did not know which could be the best name... 🤔  Anyway, let me change the var name `remote_results` to `result` because it does not make really sense

File: conans/cli/commands/list.py (hunk 6 above)
Comment (tapia) [reply conans/cli/commands/list.py]: Sorry for reopening this, but I think it would be cleaner if you check `values` only once:

```python
for prop_name, values in props.items():
    if not values:
        continue

    if prop_name in requires_fields:
        blablabla
    elif prop_name in general_fields:
        blablabla
```

File: conans/cli/commands/list.py [hunk 9]
```diff
@@ -134,68 +163,76 @@ def _get_remotes(conan_api, args):
...
+
+    return results
+
 
-    remotes = _get_remotes(conan_api, args)
+@conan_subcommand(formatters=list_package_ids_formatters)
+def list_package_ids(conan_api, parser, subparser, *args):
+    """
+    List all the package IDs for a given recipe reference. If it doesn't include the revision,
+    the command will retrieve all the package IDs for the latest one by default.
```
Comment (tapia) [inline conans/cli/commands/list.py]: A little confusing. I'd rephrase it slightly:

List all the package IDs for a given recipe reference. If the reference doesn't include the recipe revision, the command will retrieve all the package IDs for the most recent revision.

File: conans/test/integration/command_v2/list_package_ids_test.py (hunk 3 above)
Comment (tapia) [reply conans/test/integration/command_v2/list_package_ids_test.py:51]: As @franramirez688 says, this is not related to this PR. I just created [an issue](https://github.com/conan-io/conan/issues/9332) to fix this.

File: conans/test/integration/command_v2/list_package_ids_test.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+          %s:.{40}
+            settings:
+              arch=.*
+              build_type=.*
+              os=.*
+            options:
+              shared=False
+            requires:
+              pkg/0.1@user/channel:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
+        """ % repr(rrev))
```
Comment (tapia) [inline conans/test/integration/command_v2/list_package_ids_test.py:157]: Why are you building all the expected outputs with `%` formatting?

Comment (franramirez688) [reply conans/test/integration/command_v2/list_package_ids_test.py:157]: If you realize I have one line like this `%s:.{40}` so I don't want to have this `40` as a `f-string` value but a regex one

File: conans/client/api/conan_api.py [hunk 11]
```diff
@@ -268,23 +269,100 @@ def _get_revisions(self, ref, remote=None):
...
+        Get all the recipe revisions given a reference from cache or remote.
+
+        :param reference: `PackageReference` without the revision
+        :param remote: `Remote` object
+        :return: `dict` with all the results, e.g.,
+                 {"remote": "my_remote_name",  # or None
+                  "reference": "libyaml/0.2.5#80b7cbe095ac7f38844b6511e69e453a:ef93ea55bee154729e264db35ca6a16ecab77eb7",
+                  "results": [{"revision": "80b7cbe095ac7f38844b6511e69e453a",
+                               "time": "2021-07-20 00:56:25 UTC"}, ...]}
+        """
```
Comment (tapia) [inline conans/client/api/conan_api.py:294]: please reformat the example to be more easily read (Also in `get_recipe_revisions` and `get_package_ids`:

```
:return: `dict` with all the results, e.g.,
  {
    "remote": "my_remote_name",  # or None
    "reference": "libyaml/0.2.5#80b7cbe095ac7f38844b6511e69e453a:ef93ea55bee154729e264db35ca6a16ecab77eb7",
    "results": [
      {
        "revision": "80b7cbe095ac7f38844b6511e69e453a",
        "time": "2021-07-20 00:56:25 UTC"
      },
      ...
    ]
  }
```

File: conans/cli/commands/list.py (hunk 8 above)
Comment (tapia) [reply conans/cli/commands/list.py]: I'm not sure this function makes a lot of sense. Besides refactoring repeated code, what does it do? Print the header? Handle errors? Show a message when the results are empty?

I mean, if you had to write the docstring of the function, what would it say?

Comment (franramirez688) [reply conans/cli/commands/list.py]: It's only supposed to avoid repeating code so I think it makes sense to have it. Perhaps the best name is `_print_common_headers()`, what do you think @tapia ?

Comment (franramirez688) [reply conans/cli/commands/list.py]: I've renamed it as `_print_common_headers`
